drivers/watchdog/wdt_nrfx.c

```diff
@@ -65,7 +65,7 @@ static int wdt_nrf_disable(const struct device *dev)
 	err_code = nrfx_wdt_stop(&config->wdt);
 
 	if (err_code != NRFX_SUCCESS) {
-		return -ENOTSUP;
+		return -EFAULT;
```

Comment:
https://github.com/zephyrproject-rtos/zephyr/blob/8526b7d4a009164a07c98606dca90f28a5736104/include/zephyr/drivers/watchdog.h#L180-L181
`nrfx_wdt_stop()` returns `NRFX_ERROR_FORBIDDEN` when [it is not allowed to stop the watchdog](https://github.com/zephyrproject-rtos/hal_nordic/blob/13ac55b5b52c905642e9c54f069109d188aa5840/nrfx/drivers/include/nrfx_wdt.h#L270C12-L270C33), so `-EPERM` should be returned here.

Reply:
That case is handled in lines 61 and 72 - 75.
I propose to display a whole function wdt_nrf_disable().

Reply:
Logic flow is a bit different.
Whether wdt_disable() is supported or not, is handled by lines 61, 72-75 in this file.

Assumption - wdt_disable() (Zephyr API) is supported.

In such case, when this function is called before wdt_setup() (Zephyr API)
nrfx_wdt_stop() in line 65 will return NRFX_ERROR_FORBIDDEN.

This is because p_cb->stoppable is 0, unless wdt_setup() (Zephyr API) is executed.
https://github.com/zephyrproject-rtos/hal_nordic/blob/master/nrfx/drivers/src/nrfx_wdt.c#L272

Reply:
IMO using `EFAULT` for incorrect API call sequence (i.e. disable before setup) and EPERM for lack of hardware support has the advantage of differentiation between these two errors - otherwise we would have single error code for both of these issues.

Additionally, we can take a look how other vendors do it.
- STM - EPERM is returned on lack of HW support: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/watchdog/wdt_iwdg_stm32.c#L138
- several vendors - EFAULT is returned when watchdog cannot be started in given driver state:
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/watchdog/wdt_rpi_pico.c#L103
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/watchdog/wdt_cc13xx_cc26xx.c#L147
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/watchdog/wdt_xilinx_axi.c#L105

Reply:
> This is because p_cb->stoppable is 0, unless wdt_setup() (Zephyr API) is executed. https://github.com/zephyrproject-rtos/hal_nordic/blob/master/nrfx/drivers/src/nrfx_wdt.c#L272

Okay. I initially misread the way that options are passed to `nrfx_wdt_reconfigure()`. Now I see that the `NRF_WDT_BEHAVIOUR_STOP_ENABLE_MASK` flag is always used if stopping is available. So indeed, the incorrect sequence of calls is the only possibility when `NRFX_ERROR_FORBIDDEN` can be returned. But since this is not that obvious, maybe a comment like "this can only happen if wdt_nrf_setup() is not called first" could be helpful.

Reply:
I added comment as suggested.